services/appmetrics/api_test.go

```diff
@@ -0,0 +1,49 @@
+package appmetrics
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"github.com/status-im/status-go/appdatabase"
+	"github.com/status-im/status-go/appmetrics"
+	"github.com/stretchr/testify/require"
+)
+
+func setupTestDB(t *testing.T) (*appmetrics.Database, func()) {
+	tmpfile, err := ioutil.TempFile("", "appmetrics-service")
+	require.NoError(t, err)
+	db, err := appdatabase.InitializeDB(tmpfile.Name(), "appmetrics-tests")
+	require.NoError(t, err)
+	return appmetrics.NewDB(db), func() {
+		require.NoError(t, db.Close())
+		require.NoError(t, os.Remove(tmpfile.Name()))
+	}
+}
+
+func TestValidateAppMetrics(t *testing.T) {
+	db, close := setupTestDB(t)
+	defer close()
+	api := NewAPI(db, make(chan appmetrics.AppMetric, 8))
+
+	validMetrics := []appmetrics.AppMetric{appmetrics.AppMetric{
+		Event:      "go/test1",
+		Value:      json.RawMessage(`"some-string"`),
```

Comment:
Do you have a link to the client side code? I would like to check if `RawMessage` is the right type here, it makes sense if you are storing raw json in a field (i.e a `[]byte` that is already some encoded json), but from the example above, it seems you are storing plain strings

Reply:
The field is flexible enough to store any JSON shape. I used strings here for testing purposes.

Reply:
Yes, that's clear, but the question is, do you need `RawMessage` at all? Currently, as far as I can see, you are not using any of the features of it, you are getting a `RawMessage`, but converting it back to a `string`, so it makes more sense (unless I am missing something), to use a `string` in this case.
If you are just sending some stringified JSON, I don't think you need to use `RawMessage`, but I'd have to see the client side code to make sure.

Reply:
It's a raw message because it could be a string or a json object. If we set it to a string and send the value `{"hello": "world"}`, the function will reject it. 

The problem is that a JS object cannot include stringified JSON (I mentioned this on the call). If it's a string, and try to send stringified JSON, it will throw up (it was actually a string in the first iteration).

Reply:
This is the call I used to check the integration:
```clojure
(fx/defn check
  {:events [::check]}
  [_]
  {::json-rpc/call [{:method "appmetrics_validateAppMetrics"
                     :params [[{:event "go/test1"
                                :val "some-string" ;; could also be a map like {:a "b" :c {:d "e"}}
                                :app_version "1.12"
                                :os "android"}]]
                     :on-success prn
                     :on-error prn}]})
```

Reply:
Oh I see, so you want to have a variable type there, I thought you were always sending stringified JSON or a string.

Ok, so if you are always sending stringified JSON, you can almost certainly use a string, you client code should be:

`:val (if (string? a) a (js/JSON.stringify a))`

That will not throw an error in status-go (my understanding is that you will not parse it)

If though you actually want to send raw-object, then yes, you need to use `RawMessage` to prevent go to parse it.

Reply:
okay, let me try this out.

Reply:
- `string` instead of `json.RawMessage` works:

<img width="1220" alt="Screenshot 2021-03-17 at 12 59 03 PM" src="https://user-images.githubusercontent.com/1925158/111430593-95641700-8720-11eb-8375-dc103877fae6.png">

- it didn't work the last time I tried because I was adding newline characters in stringified JSON and Insomnia probably did some magic before sending the request
- I can change it back to string, but is there any extra cost of using `json.RawMessage`? Because sending pure json makes it easy to test via Insomnia (compared to strings, which are fragile and hard to manage)

Reply:
It's a bit less obvious what's happening when using `RawMessage` but I don't mind, maybe leave a comment on the field so it's clear (i.e we use `RawMessage` so that we can send any type of data and not just strings)

Reply:
I'll add a comment.